cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/KeyHelper.java

```diff
@@ -77,42 +84,42 @@ static void addTargetIdToKey(MDSKey.Builder builder, Id.NamespacedId namespacedI
     }
   }
 
-  static Id.NamespacedId getTargetIdIdFromKey(MDSKey.Splitter keySplitter, String type) {
-    if (type.equals(Id.Program.class.getSimpleName())) {
+  static NamespacedId getTargetIdIdFromKey(MDSKey.Splitter keySplitter, String type) {
+    if (type.equals(ProgramId.class.getSimpleName())) {
       String namespaceId = keySplitter.getString();
       String appId = keySplitter.getString();
       String programType = keySplitter.getString();
       String programId = keySplitter.getString();
-      return Id.Program.from(namespaceId, appId, ProgramType.valueOf(programType), programId);
-    } else if (type.equals(Id.Application.class.getSimpleName())) {
+      return ProgramId.fromIdParts(Arrays.asList(namespaceId, appId, programType, programId));
```

Comment:
instead of this, why not use ``new ProgramId(...)``? applies to this entire method